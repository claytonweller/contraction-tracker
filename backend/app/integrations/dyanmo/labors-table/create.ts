import * as AWS from 'aws-sdk'

const dynamo = new AWS.DynamoDB.DocumentClient()
export async function createLabor(newLabor) {
  const getParams = {
    TableName: 'laborsTable',
    Key: {
      userId: "123"
    }
  }
  const existingLabors = await dynamo.get(getParams).promise()

  const labor = {
    userId: "123",
    labors: [...existingLabors?.Item?.labors, newLabor]
  }
  const laborInfo = {
    TableName: 'laborsTable',
    Item: labor,
  };
  const result = await dynamo.put(laborInfo).promise()
  console.warn(result)
  return result
}