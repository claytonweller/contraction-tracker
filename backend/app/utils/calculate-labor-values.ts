import { ILabor } from "../types/Labor"
import { ICalculatedLabor } from "../types/CalculatedLabor"
import { DateTime } from "luxon"
import { IContraction } from "../types/Contraction"

export function calculateLaborValues(labor: ILabor): ILabor {
  const contraction = calculateContractionValues(labor)
  const rest = calculateRestValues(labor)
  return {
    ...labor,
    calculated: {
      contraction,
      rest
    }
  }
}

function calculateContractionValues(labor: ILabor): ICalculatedLabor['contraction'] {
  const { contractions } = labor

  const { lastHour: lastHourIntensities, all: intensities } = extractIntesities(contractions)
  const { lastHour: lastHourDurations, all: durations } = extractDurations(contractions)
  const averageDuration = calculateAverage(lastHourDurations)
  const averageIntensity = calculateAverage(lastHourIntensities)

  return {
    durations,
    intensities,
    averageDuration,
    averageIntensity
  }
}

function extractIntesities(contractions: IContraction[]) {
  return contractions.reduce((calculated, contraction): any => {
    const { intensity, endTime } = contraction
    if (!endTime || !intensity) return calculated
    return addToAllAndLastHour(calculated, intensity, contraction)
  }, { lastHour: [], all: [] })
}

function extractDurations(contractions: IContraction[]) {
  return contractions.reduce((calculated, contraction): any => {
    const { endTime, startTime } = contraction
    if (!endTime) return calculated
    const duration = calculateDuration(startTime, endTime)
    if (duration <= 0) return calculated
    return addToAllAndLastHour(calculated, duration, contraction)
  }, { lastHour: [], all: [] })
}}

function addToAllAndLastHour(
  calculated: { all: number[], lastHour: number[] },
  newValue: number,
  contraction: IContraction
) {
  const all = [...calculated.all, newValue]
  const lastHourDurations = isInLastHour(contraction)
    ? [...calculated.lastHour, newValue]
    : calculated.lastHour
  return { all, lastHourDurations }
}

function isInLastHour(contraction: IContraction): boolean {
  const start = DateTime.fromISO(contraction.startTime)
  return DateTime.now().diff(start, 'minutes').minutes < 60
}

function calculateAverage(arr: number[]): number {
  const total = arr.reduce((sum, duration) => sum + duration, 0)
  return total / arr.length
}

function calculateDuration(startTime: string, endTime?: string): number {
  if (!endTime) return 0
  const start = DateTime.fromISO(startTime)
  const end = DateTime.fromISO(endTime)
  return end.diff(start, 'seconds').seconds
}

function calculateRestValues(labor: ILabor): ICalculatedLabor['rest'] {
  return {
    current: 1,
    average: 1,
    durations: [1]
  }
}