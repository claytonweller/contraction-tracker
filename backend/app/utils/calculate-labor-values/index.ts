import { ILabor } from "../../types/Labor"
import { ICalculatedLabor } from "../../types/CalculatedLabor"
import { calculateContractionValues } from "./calculate-contraction-values"

export function calculateLaborValues(labor: ILabor): ILabor {
  const contraction = calculateContractionValues(labor)
  const rest = calculateRestValues(labor)
  return {
    ...labor,
    calculated: {
      contraction,
      rest
    }
  }
}


function calculateRestValues(labor: ILabor): ICalculatedLabor['rest'] {
  return {
    current: 1,
    average: 1,
    durations: [1]
  }
}