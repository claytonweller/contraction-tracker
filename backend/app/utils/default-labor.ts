import { DateTime } from "luxon";
import { IContraction } from "../types/Contraction";
import { ILabor } from "../types/Labor";

export const defaultLabor = (params:{
  totalMinutesOfLabor?: number,
  contractions?: IContraction[],
}={}): ILabor => {
  const now = DateTime.now()
  const {
    totalMinutesOfLabor = 30,
    contractions = [
      {
        startTime: now.minus({ minute: 25 }).toISO(),
        endTime: now.minus({ minute: 24 }).toISO(),
        intensity: 1
      },
      {
        startTime: now.minus({ minute: 20 }).toISO(),
        endTime: now.minus({ minute: 18 }).toISO(),
        intensity: 2
      }

    ]
  } = params
  
  return {
    userId: "1",
    startTime: now.minus({ minute: totalMinutesOfLabor }).toISO(),
    endTime: undefined,
    contractions,
    calculated: {
      isGoTime: false,
      contraction: {
        averageDuration: 6000,
        averageIntensity: 1,
        durations: [],
        intensities: []
      },
      rest: {
        currentDuration: 6000,
        averageDuration: 6000,
        durations: [],
      }
    },
    bishopScore: 0
  }
}

