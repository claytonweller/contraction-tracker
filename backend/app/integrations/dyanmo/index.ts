import * as AWS from 'aws-sdk'
import { connectToLaborTable } from "./labors-table";
import { connectToUserTable } from "./users-table";

const client = new AWS.DynamoDB.DocumentClient()

export const dynamo = {
  labor: connectToLaborTable(client),
  user: connectToUserTable(client)
}

export interface IBaseDynamoParams {
  TableName: string
}