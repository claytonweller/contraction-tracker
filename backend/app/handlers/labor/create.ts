import { withErrorHandling } from "../../utils/with-error-handling"
import { dynamo } from "../../integrations/dyanmo"
import { calculateLaborValues } from "../../utils/calculate-labor-values";

async function createHandler(event) {
  console.log('CREATE', event)
  const labor = JSON.parse(event.body)
  const laborWithCalculatedValues = calculateLaborValues(labor)
  const createdLabor = await dynamo.labor.create(laborWithCalculatedValues, "123")
  return createdLabor
}

export const handler = withErrorHandling(createHandler)
