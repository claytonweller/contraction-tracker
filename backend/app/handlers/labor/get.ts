import { dynamo } from "../../integrations/dyanmo"
import { withErrorHandling } from "../../utils/with-error-handling"
import {z} from 'zod'

export async function getHandler(event) {
  const {userId} = event.queryStringParameters
  const validUserId = z.string().parse(userId)
  const labors = await dynamo.labor.get(validUserId)
  return labors
}

export const handler = withErrorHandling(getHandler)