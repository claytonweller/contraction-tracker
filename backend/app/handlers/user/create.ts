import { dynamo } from "../../integrations/dyanmo"
import { withErrorHandling } from "../../utils/with-error-handling"

export async function createHandler(event) {
  const params = JSON.parse(event.body)
  const createdUser = await dynamo.user.create(params)
  return createdUser
}

export const handler = withErrorHandling(createHandler)