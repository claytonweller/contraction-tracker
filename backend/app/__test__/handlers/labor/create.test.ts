import { createHandler as createLaborHandler } from "../../../handlers/labor/create";
import { dynamo } from "../../../integrations/dyanmo";
import * as calculateLaborValues from "../../../utils/calculate-labor-values";
import { defaultLabor } from "../../../utils/default-labor";


describe('createLaborHandler', ()=>{
  const dynamoSpy = jest.spyOn(dynamo.labor, 'create')
  const calculateSpy = jest.spyOn(calculateLaborValues, 'calculateLaborValues')
  beforeEach(()=>{
    jest.clearAllMocks()
    calculateSpy.mockImplementation((input)=> input)
  })
  const validLabor = {...defaultLabor(), endTime: undefined}

  describe('when the body is incorrectly formatted', ()=>{
    const incorrect = {body:JSON.stringify({startTime:'123'})}
    let error 
    beforeEach(async ()=>{
      try {  
        await createLaborHandler(incorrect)
      } catch (e){
        error = e
      }
    })  
    
    it('should throw an error', ()=>{
      expect(error).toMatchSnapshot()
    })

    it('should not attempt to persist the labor', ()=>{
      expect(dynamoSpy).not.toHaveBeenCalled()
    })
  })

  describe('when the body is correctly formatted', ()=>{
    const correct = {body:JSON.stringify(validLabor)}

    describe('and the labor is successfully persisted', ()=>{
      let result
      beforeEach(async()=>{
        dynamoSpy.mockResolvedValue(validLabor)
        result = await createLaborHandler(correct)
      })
      it('should return a correctly formatted labor', ()=>{
        expect(result).toEqual(validLabor)
      })
      it('calculates the labor values', ()=>{
        expect(calculateSpy).toHaveBeenCalledWith(validLabor)
      })
      it('should have attempted to persist the labor', ()=>{
        expect(dynamoSpy).toHaveBeenCalledWith(validLabor)
      })
    })

    describe('and the labor is NOT successfully persisted', ()=>{
      const message = 'KABOOM!'
      let error 
      beforeEach(async ()=>{
        dynamoSpy.mockRejectedValue(message)
        try {  
          await createLaborHandler(correct)
        } catch (e){
          error = e
        }
      })  
      it('should throw an error', ()=>{
        expect(error).toBe(message)
      })
      it('calculates the labor values', ()=>{
        expect(calculateSpy).toHaveBeenCalledWith(validLabor)
      })
      it('should have attempted to persist the labor', ()=>{
        expect(dynamoSpy).toHaveBeenCalledWith(validLabor)
      })
    })
  })
})