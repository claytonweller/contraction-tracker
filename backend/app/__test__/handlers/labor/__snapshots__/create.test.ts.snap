// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLaborHandler when the body is incorrectly formatted should throw an error 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "userId"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "contractions"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "calculated"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "bishopScore"
    ],
    "message": "Required"
  }
]]
`;
